[Response] `getMaxAge()` returns non-negative integer

pkruithof opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Refs https://github.com/symfony/symfony/pull/48651#issuecomment-1371079091
| License       | MIT
| Doc PR        | 

The `max-age` directive should be a non-negative integer, see [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control):

> The max-age=N request directive indicates that the client allows a stored response that is generated on the origin server within N seconds — where N may be any non-negative integer (including 0).

In case the value is negative, it's encouraged to be treated as 0:

> In other words, for any max-age value that isn't an integer or isn't non-negative, the caching behavior that's encouraged is to treat the value as if it were 0.

In my case, it lead to a response that was `private,no-cache` but with an `Expires` header set in the future. Not every browser handled this inconsistency the same, which eventually led to authentication issues (see linked comment for a more elaborate explanation).

pkruithof committed:
[Response] `getMaxAge()` returns non-negative integer

pkruithof commented on src/Symfony/Component/HttpFoundation/Tests/ResponseTest.php:
```diff
@@ -353,9 +353,8 @@ public function testGetMaxAge()
-        $response->headers->set('Cache-Control', 'must-revalidate');
         $response->headers->set('Expires', -1);
-        $this->assertLessThanOrEqual(time() - 2 * 86400, $response->getExpires()->format('U'));
+        $this->assertSame(0, $response->getMaxAge());
```
The test was for `getMaxAge()` however in this case it was testing `getExpires()`, which seemed wrong.

pkruithof commented:
There is a failing test for `ResponseTest::testGetTtl()` where the TTL is asserted to be negative when the response's expiration is in the past. This uses the now changed `getMaxAge()`. However I'm not sure if I should change the test to assert `0` instead of a negative number, or if `getTtl()` should be changed to be compliant with the test.

It seems to me that a negative TTL is illogical the same way as a negative `max-age`. Please advise what to do with this.

fabpot commented:
> There is a failing test for `ResponseTest::testGetTtl()` where the TTL is asserted to be negative when the response's expiration is in the past. This uses the now changed `getMaxAge()`. However I'm not sure if I should change the test to assert `0` instead of a negative number, or if `getTtl()` should be changed to be compliant with the test.
> 
> It seems to me that a negative TTL is illogical the same way as a negative `max-age`. Please advise what to do with this.

I supposeI would expect `0`.

pkruithof commented:
> > There is a failing test for `ResponseTest::testGetTtl()` where the TTL is asserted to be negative when the response's expiration is in the past. This uses the now changed `getMaxAge()`. However I'm not sure if I should change the test to assert `0` instead of a negative number, or if `getTtl()` should be changed to be compliant with the test.
> > It seems to me that a negative TTL is illogical the same way as a negative `max-age`. Please advise what to do with this.
> 
> I supposeI would expect `0`.

Agreed, I've changed the `getTtl()` method accordingly.

fabpot commented on src/Symfony/Component/HttpKernel/Tests/EventListener/SessionListenerTest.php:
```diff
@@ -590,6 +590,30 @@ public function testResponseHeadersMaxAgeAndExpiresDefaultValuesIfSessionStarted
+
+        $session = $this->createMock(Session::class);
+        $session->expects($this->once())->method('getUsageIndex')->willReturn(1);
+        $request = new Request(server: ['SERVER_PROTOCOL' => 'HTTP/1.0']);
```
This should be changed to be compatible with all supported PHP versions (see failing CI)

fabpot reviewed (approved):
All failing tests seem unrelated to your changes.

fabpot commented:
Went too fast. This one seems relate (on low-deps):

```
1) Symfony\Component\HttpKernel\Tests\EventListener\SessionListenerTest::testPrivateResponseMaxAgeIsRespectedIfSessionStarted
Failed asserting that -172800 is identical to 0.

/home/runner/work/symfony/symfony/src/Symfony/Component/HttpKernel/Tests/EventListener/SessionListenerTest.php:609
```

pkruithof commented:
I noticed that too, the only way that should be possible is if the `s-maxage` or `max-age` directives are negative. I want to debug this but I cannot reproduce the failing test. Here's what I did:

```
docker run --rm -it -v $PWD:/symfony php:8.1-alpine sh
# install xsl extension, composer and cd to `/symfony`
/symfony # export COMPOSER_ROOT_VERSION=5.4.x-dev
/symfony # export SYMFONY_VERSION=5.4
/symfony # composer update --prefer-lowest
```

This gave the following problem: 
```
    - Root composer.json requires symfony/runtime 5.4.x-dev, it is satisfiable by symfony/runtime[5.4.x-dev] from composer repo (https://repo.packagist.org) but symfony/runtime[dev-main] from path repo (src/Symfony/Component/Runtime) has higher repository priority. The packages from the higher priority repository do not match your constraint and are therefore not installable. That repository is canonical so the lower priority repo's packages are not installable. See https://getcomposer.org/repoprio for details and assistance.
```

To get around this I temporarily removed the custom repo path for the runtime component (it is not related in any way to this change). Then it did install the dependencies, and the test runs without issues:

```
/symfony # ./phpunit --filter SessionListenerTest

PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

Testing 
................................................                  48 / 48 (100%)

Time: 00:03.350, Memory: 295.00 MB

OK (48 tests, 226 assertions)
```

So I'm not sure what to do now?

pkruithof commented:
@fabpot could you point me in the right direction to move this PR forward?

lyrixx commented:
Wohoo. Thanks @pkruithof I hit the very same bug (not so easy to find). Your bug report is excellent. I tested your PR and it fixes my issue to.

fabpot commented:
@pkruithof The problem is that your change impacts 2 components (which are independent). To fix the issue, you should bump the minimum version of HttpFoundation in the HttpKernel composer.json file. Right now, it's `^5.4|^6.0`, this should be bumped to the next patch versions of 5.4, 6.0, 6.1, and 6.2. That will fix the tests.

fabpot commented on src/Symfony/Component/HttpKernel/composer.json:
```diff
@@ -20,7 +20,7 @@
         "symfony/error-handler": "^4.4|^5.0|^6.0",
         "symfony/event-dispatcher": "^5.0|^6.0",
-        "symfony/http-foundation": "^5.3.7|^6.0",
+        "symfony/http-foundation": "^5.4|^6.0|^6.1|^6.2",
```
You must specify the patch versions here. For 5.4 for instance, that would be 5.4.21 as this is going to be the next version.

pkruithof replied:
Where can I find which patch versions are needed? I'm not really familiar with this part so I'm a bit lost what the correct solution should be here. Or is it just the next patch version for all release branches?

fabpot replied:
You can have a look at the tags and add +1. I've done it for you. So, everything is ok now.
Note that I've removed 6.0 and 6.1 here as they are not maintained anymore and as such will never have this patch.

pkruithof replied:
Thank you

fabpot commented:
Thank you @pkruithof.
